generative/cxxbridge/noder.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+#include "generative/generative/cxxbridge/coord_ffi.rs.h"
+#include "generative/generative/cxxbridge/geometry_graph_ffi.rs.h"
+#include "geometry_collection.hpp"
+#include "geometry_graph.hpp"
+#include <generative/noding/geometry-graph.h>
+#include <generative/noding/geometry-noder.h>
+
+#include <geos/geom/Geometry.h>
+#include <geos/noding/snap/SnappingNoder.h>
+#include <geos/operation/overlay/snap/GeometrySnapper.h>
+#include <geos/operation/polygonize/Polygonizer.h>
+
+#include <iterator>
+#include <memory>
+
+[[nodiscard]] inline std::unique_ptr<GeometryGraphShim>
+node(const GeometryCollectionShim& rust_geoms, double tolerance) noexcept
+{
+    const auto geos_geoms = copy_rust_collection_to_geos(rust_geoms);
+    auto noded = std::unique_ptr<geos::geom::Geometry>(nullptr);
+    if (tolerance == 0.0)
+    {
+        noded = generative::noding::GeometryNoder::node(*geos_geoms, nullptr);
+    } else
+    {
+        auto noder = std::make_unique<geos::noding::snap::SnappingNoder>(tolerance);
+        noded = generative::noding::GeometryNoder::node(*geos_geoms, std::move(noder));
```

Comment:
The `IteratedNoder` can fail to converge, and throw a TopologyException if it does so

![image](https://github.com/Notgnoshi/generative/assets/8878460/7a595fc2-6922-4a0a-8180-a7fbc0d39414)

Increasing the number of iterations didn't work :sob:

There's two options:
1. Use the SnappingNoder instead of the IteratedNoder in these cases
   1. Choose the noder with a CLI argument
   2. Don't use the Rust k-D Tree snapping, and use the SnappingNoder whenever --snap is given
3. Catch the error, don't crash, but also don't resolve
4. Catch the error, and retry with the SnappingNoder

Always using the SnappingNoder sucks because then you lose any isolated POINTs that aren't attached to a segment when noding.

It _is_ possible to detect any points that aren't in the noded graph and resolve, _except_ that if those points are on top of a segment, that segment will not be noded.

Given all these cons, and an utter lack of pros, I think there should be a --snapping-noder CLI option